pkgs/build-support/trivial-builders/default.nix

```diff
@@ -127,7 +127,14 @@ rec {
 
   # See doc/build-helpers/trivial-build-helpers.chapter.md
   # or https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-text-writing
-  writeText = name: text: writeTextFile { inherit name text; };
+  writeText = name: text: writeTextFile {
+    inherit name text;
+    derivationArgs = rec {
+      outputHashAlgo = "sha256";
+      outputHashMode = "flat";
+      outputHash = builtins.hashString outputHashAlgo (builtins.toString text);
```

Comment:
This is an interesting idea.  I don't think I've seen this type of thing tried before.

I couldn't come up with a reason not to do this, except maybe:

- Does having `writeText` not be fixed output cause a lot of rebuilds in practice?  It feels like it should be relatively trivial (at least compared to everything else Hydra has to do), but I don't really have a good intuition for this.

- Would this change cause any significant amount of increased evaluation time?  (As long as the hashes for all calls to `writeText` aren't calculated at evaluation time, it seems like evaluation time likely wouldn't significantly increase?)

Reply:
I have two issues:

* Aren't all inputs to `writeText` already pure? Do rebuilds really happen without FODs?
* If FODs are necessary to avoid rebuilds, it might be better to create a separate FOD variant. `writeText` is often used in NixOS and Home Manager modules. Converting all uses of `writeText` to FOD would require hashes to be specified everywhere throughout NixOS and Home Manager configuration.

Reply:
You don't need to specify any hashes since we can hash everything without causing IFD or side effects. 

Name shouldn't change the hash of the FOD, can it? 

But tbh, this should be a ca-derivativion 😂

Reply:
> Aren't all inputs to writeText already pure? Do rebuilds really happen without FODs?

`writeText` just calls `writeTextFile` under the hood, which uses `runCommand` (which uses `stdenv.mkDerivation`, right?).  So calls to `writeText` will rebuild whenever something in `runCommand` / `stdenv.mkDerivation` changes, right?

> Converting all uses of writeText to FOD would require hashes to be specified everywhere throughout NixOS and Home Manager configuration.

Like Sandro said, it is not necessary to manually specify hashes in calls to `writeText`, since the input text is available in Nix, and then the `builtins.hashString` function is used to automatically calculate the hash.

> Name shouldn't change the hash of the FOD, can it?

I don't the the hash changes if the name changes.  However, I believe that the derivation will get rebuilt if the name changes, even if the hash doesn't change.  (I talk about this a little bit in https://functor.tokyo/blog/2023-07-24-fodonuts).

Reply:
It should definitely go into staging thanks--updated.

@cdepillabout is correct about current (non-FOD) writeText causing plenty rebuilds. Example:

```
$ nix-build --no-link -E '(import <nixpkgs> {}).writeText "foo" "foobar"'
/nix/store/plm84lpdixks9wfh5a4pp2k5sgzsmfxd-foo
$ nix-build --no-link -E '(import <nixpkgs> { overlays = [(self: super: { gnutar = super.gnutar.overrideAttrs { name = "foobar"; };}) ];}).writeText "foo" "foobar"'
/nix/store/wwsbs9xckbrmj7zvwxlqcj3s4393gp45-foo
```

Yes the name change will trigger a rebuild (because the $out will change) but that's ok--this patch is more any of the *upstream* dependencies of writeText itself causing a rebuild. In the case of writeText this is pointless because we know the exact contents of the output in advance. On larger build farms this can cause an avalanche of rebuilds for no particular reason.

> If FODs are necessary to avoid rebuilds, it might be better to create a separate FOD variant. writeText is often used in NixOS and Home Manager modules. Converting all uses of writeText to FOD would require hashes to be specified everywhere throughout NixOS and Home Manager configuration.

@SuperSandro2000 already answered this, but this patch is a drop-in replacement for writeText, itself computing the hash from its own input. The text is known at runtime so you can just calculate the output hash during eval.

> But tbh, this should be a ca-derivativion 😂
I also thought about this, and am on the fence. I guess the trade-off is: doing FOD like this will increase all eval time, but is strictly more powerful for build farms who never even have to try rebuilding. Doing CA keeps eval time lower but does require that initial rebuild of only the text file itself every time something changes, just to realize that nothing has changed in the output.

Maybe something to measure?

Reply:
Finally:

> Does having writeText not be fixed output cause a lot of rebuilds in practice? It feels like it should be relatively trivial (at least compared to everything else Hydra has to do), but I don't really have a good intuition for this.

I don't have any hard numbers but I manage a decent sized scope ("https://github.com/hraban/cl-nix-lite") with a few hundred packages which get rebuilt nightly based on nixpkgs-unstable: take this for the hearsay it is, but I feel like writeText is particularly painful because it blocks an entire "worker" on a menial non-CPU intensive task for a while. This means even with something like `-j 30`, I spend a few minutes at the start of each build at ~5% CPU just waiting for all these carbon copy files to be recreated before it can move on to actual building.

Obviously this is just based off of eye-balling build outputs, nothing scientific. Is there a way to just "try" a full hydra rebuild and compare numbers?

Reply:
Cool, I didn't realize FODs could be used this way.

> Obviously this is just based off of eye-balling build outputs, nothing scientific. Is there a way to just "try" a full hydra rebuild and compare numbers?

I tried something like this:

```console
❯ nix-build --dry-run -E 'let pkgs = import ./. { overlays = [ (final: prev: { gnutar = prev.gnutar.overrideAttrs { name = "foobar"; }; }) ]; }; in pkgs.mpv'
these 1554 derivations will be built
...
these 758 paths will be fetched (2908.90 MiB download, 9389.42 MiB unpacked):
...
```

But got the same numbers before and after the change. Maybe there are better examples other than `pkgs.mpv`.